// react-s3-typescript/dist/esm/react-aws-s3.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import shortId from 'short-uuid';
import { throwUploadError } from './ErrorThrower';
import GetUrl from './Url';
import { S3Client, PutObjectCommand, DeleteObjectCommand } from "@aws-sdk/client-s3";
class ReactS3Client {
    constructor(config) {
        this.config = config;
    }
    uploadFile(file, newFileName) {
        return __awaiter(this, void 0, void 0, function* () {
            throwUploadError(this.config, file);
            let fileExtension = '';
            if (file.name) {
                fileExtension = file.name.split('.').pop() || '';
            }
            if (!fileExtension && file.type != null) {
                fileExtension = file.type.split('/').pop() || '';
            }
            const fileName = `${newFileName || shortId.generate()}${fileExtension && '.' + fileExtension}`;
            const dirName = (this.config.dirName ? this.config.dirName + '/' : '').replace(/([^:]\/)\/+/g, '$1');
            const key = `${dirName}${fileName}`;
            const url = GetUrl(this.config);
            const client = new S3Client({
                region: this.config.region,
                credentials: {
                    accessKeyId: this.config.accessKeyId,
                    secretAccessKey: this.config.secretAccessKey,
                }
            });
            const params = {
                Bucket: this.config.bucketName,
                Key: key,
                Body: file,
                ACL: "public-read",
                Metadata: {
                    uuid: "14365123651274",
                    tag: "",
                },
                ContentType: file.type,
                ServerSideEncryption: "AES256",
            };
            const command = new PutObjectCommand(params);
            const data = yield client.send(command);
            if (data["$metadata"]["httpStatusCode"] !== 200)
                return Promise.reject(data);
            return Promise.resolve({
                bucket: this.config.bucketName,
                key,
                location: `${url}/${key}`,
                status: 0,
            });
        });
    }
    deleteFile(key) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = new S3Client({
                region: this.config.region,
                credentials: {
                    accessKeyId: this.config.accessKeyId,
                    secretAccessKey: this.config.secretAccessKey,
                }
            });
            const params = {
                Bucket: this.config.bucketName,
                Key: key,
            };
            const deleteCommand = new DeleteObjectCommand(params);
            return new Promise((resolve, reject) => {
                client.send(deleteCommand, (err) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve({
                            message: "File Deleted Successfully!",
                            key: key
                        });
                    }
                });
            });
        });
    }
}
export default ReactS3Client;
